VTS documentation update (#85)
onmax/nimiq-developer-center: learn/validator-trustscore.md

# Nimiq Validator Trustscore

<Callout type='warning'>

>  The Validator Trustscore is still under heavy development and nothing is final. Feel free to share your [Suggestions and Feedback](#suggestions-feedback).

Reviewer: ```suggestion
The Validator Trustscore is still under development and we are open to discuss changes to the algorithm.
```

The current copy feels like we are currently changing the algorithm, which is not the case. Maybe we should say that we are open to changing it if we think the feedback is useful?

</Callout>

> The Validator Trustscore (VTS) helps users assess the reliability of validators in the Nimiq Wallet. This score ranges from 0 to 1, where 0 indicates a validator is not trustworthy, and 1 indicates a highly trustworthy validator. The VTS enables stakers make informed decisions when selecting validators.

Reviewer 2: ```suggestion
The Validator Trustscore (VTS) helps users assess the reliability of validators in the Nimiq Wallet. This score ranges from 0 to 1, where 0 indicates a validator is not trustworthy, and 1 indicates a highly trustworthy validator. The VTS enables stakers to make informed decisions when selecting validators.
```

</Callout>

### Dominance

The dominance factor ensures that no single validator controls an excessive portion of the network's total stake. Validators with a larger stake receive a lower score to encourage a fairer distribution of control across the network. This approach penalizes disproportionately large stakes while promoting diversity in validator participation.

#### Dominance Ratio

The dominance ratio ($s$) of a validator is calculated using one of two methods, depending on the epoch's state. However, due to technical limitations, only the **Active Epoch Method** is currently implemented and used.

**Active Epoch Method**:
During an active epoch, the dominance ratio is determined by dividing the validator's share by the total network share:

$$
s = \frac{v}{Z}
$$

Where:
- $v$: The validator's share.
- $Z$: The total network share.

This method relies on the `getActiveValidators` function from the [RPC](https://www.nimiq.com/developers/build/set-up-your-own-node/rpc-docs/#getactivevalidators), which provides real-time balances of each active validator.

> **Closed Epoch Method**:

Reviewer: ```suggestion
**Finished Epoch Method**:
```

What do you think about using "finished" instead?

> For a closed epoch, a less precise fallback method exists, but is not currently implemented. This approach calculates the dominance ratio based on slot distribution across voting blocks. The ratio is derived by dividing the number of slots allocated to a validator by the total slots in the epoch:

Reviewer: Actually, it is implemented. In the code we have:
- `dominanceRatioViaBalance`: This value is exact. The only drawback is that we can only fetch this value when the epoch is active
- `dominanceRatioViaRatio`: This value is an approximation and not exact. This approximation is really close

Here is a screenshot of the values we have in our database:

![image](https://github.com/user-attachments/assets/5b03da93-0790-420a-ac4b-44523a2d1f74)

Reviewer: I can update this section of the docs if you don't have time or you are not sure. Let me know :)

$$
s = \frac{sl}{Sl}
$$

Where:

- $sl$: The number of slots allocated to the validator.
- $Sl$: The total number of slots in the epoch.

While this method is defined in the code as `dominanceRatioViaSlots`, it is not actively used in practice.

#### Curve adjustment

After determining the dominance ratio, a curve is applied to calculate the dominance score $S$:

$$
S = \max \left( 0 , 1 - s^{k} \right), \quad \text{being } t = 0.15 \text{ and } k=7.5
$$

Where:

> - $t = 0.25$ is the threshold.

Reviewer 2: ```suggestion
- $t = 0.15$ is the threshold.
```

Author: ![image](https://github.com/user-attachments/assets/78fb9162-2450-4233-8833-d1c23c05b3f1)
@onmax send me these updated values in the Hackathon
(?)

Reviewer 2: This is the one: https://www.desmos.com/calculator/xynm6uphlq

> - $k = 4$ is the slope of the curve.

Reviewer 2: ```suggestion
- $k = 7.5$ is the slope of the curve.
```

| 20%              | 0.59        |
| >=25%            | 0           |

### Reliability

The Reliability factor measures how consistently a validator produces blocks when expected. Validators who reliably produce their assigned blocks will have a high score, while those who fail frequently will receive a lower score. The reliability score is calculated as a weighted moving average over multiple epochs, emphasizing recent performance.

> #### Calculating Reliability for an Epoch

Reviewer: ```suggestion
#### Calculating Reliability for an Epoch

```

missing an empty line in between

<figcaption>

Graph of the reliability score adjustment. The x-axis represents the reliability score, and the y-axis represents the adjusted reliability score. The adjustment curve is derived from a circular arc with a defined center.

</figcaption>

</figure>

<br/>

For example, a baseline reliability score of $\bar{R}=0.9$ may correspond approximately to 10% downtime, reflecting significant but not catastrophic validator underperformance. Validators with lower scores experience increasingly steep penalties, ensuring that unreliable validators are disincentivized.

Note that using 10% is only a heavy approximation. The value of 0.9 could represent 10% downtime, but also 20% or 5%, depending on when the downtime occurred. We use 10% to provide a relatable scale for the reader.

### Availability

The availability factor measures how often a validator is online and selected to produce blocks. Validators that are consistently selected and actively producing blocks receive a higher score, promoting reliable participation and network security.

#### Why Availability Matters

> Without availability, a validator could have high dominance and reliability scores but still not actively contribute to the network. This would misrepresent their role in maintaining the network's operation. The availability factor ensures only validators actively selected to produce blocks receive a higher score, penalizing those that are inactive, jailed, or offline.

Reviewer 2: Here you are mentioning offline status, but in the next paragraph you say:
```
Availability focuses on participation in block production, regardless of online or offline status.
```
